### PR TITLE
Support IN16B's single detectors in IndirectILLEnergyTransfer

### PR DESCRIPTION
**Pull request by MathieuTillet**
**Description of work.**
In BATS mode, IN16B can have single detectors, that needs the same treatments as the other detectors, but without grouping.
This PR takes care of this, and fixes an indexation bug for regular pixel grouping.

**To test:**
- Run
 ```
center, epp = IndirectILLEnergyTransfer(Run='215962', PulseChopper='34', GroupDetectors=False)

offset_p100 = IndirectILLEnergyTransfer(Run='215967', InputElasticChannelWorkspace=epp, PulseChopper='34', GroupDetectors=False)
```
(with data from `ExternalData/Testing/SystemTest/ILL/IN16B`)
- Plot the color fill of the results and check there is an elastic peak neatly aligned.
- Plot spectra 2050 and 2051, and check the elastic peak is correctly centered on 0. 

WIP for #30591 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py [hunk 1]
```diff
@@ -437,10 +442,14 @@ def _create_elastic_channel_ws(self, epp_ws, run, epp_equator_ws=None):
         epp_ws.addColumn('double', 'ChopperSpeed')
         epp_ws.addColumn('double', 'ChopperPhase')
         epp_ws.addColumn('double', 'PSD_TOF_Delay')
+        equator_epp = 1 if epp_equator_ws else 0
         for row in range(epp_ws.rowCount()):
             epp_ws.setCell('ChopperSpeed', row, speed)
             epp_ws.setCell('ChopperPhase', row, phase)
-            epp_ws.setCell('PSD_TOF_Delay', row, delay)
+            if epp_ws.rowCount() - single_detectors - equator_epp <= row < epp_ws.rowCount() - equator_epp:
+                epp_ws.setCell('PSD_TOF_Delay', row, delay_sd)
```
Comment (gvardany) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py:450]: Now the delay is both for PSD and SD, so let's rename it to just `TOF_Delay`.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py [hunk 2]
```diff
@@ -502,26 +512,44 @@ def _convert_to_energy_bats(self, ws, epp_ws, t0_offset=0.):
...
+        for single_detector in range(N_PIXELS_PER_TUBE * N_TUBES + N_MONITOR, ws.getNumberHistograms()):
+            group = rows - 1 - (ws.getNumberHistograms() - single_detector)
+            if epp_ws.cell('FitStatus', group) in ['success', "narrowPeak"]:
+                l2 = detector_info.l2(single_detector)
+                elastic_tof = ((l1 + l2) / v_fixed + sd_t0_offset) * 1E+6
+                elastic_channel = epp_ws.cell('PeakCentre', group)
+            else:
+                elastic_tof = elastic_tof_equator
+                elastic_channel = elastic_channel_equator
+                self.log.warning("Single detector number {0}'s peak was not found. Using equatorial peak instead."
```
Comment (gvardany) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py:547]: In this case could we not calculate a guess based on the relative distance and relative delay differences of the PSD equator and SD?

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py [hunk 3]
```diff
@@ -592,26 +621,48 @@ def _reduce_bats(self, ws):
...
+            psd_t0_offset = self._t0_offset(center_chopper_speed, center_chopper_phase,
+                                            shifted_chopper_phase, center_psd_delay, shifted_psd_delay)
+            sd_t0_offset = 0
+            if single_detectors:
+                shifted_sd_delay = run.getLogData('SingleD.time_of_flight_2').value
+                center_sd_delay = input_epp.cell('PSD_TOF_Delay', input_epp.rowCount() - 2)
+                sd_t0_offset = self._t0_offset(center_chopper_speed, center_chopper_phase, shifted_chopper_phase,
+                                               center_sd_delay, shifted_sd_delay)
+
+            self.log().information('T0 Offset is {0} [sec]'.format(psd_t0_offset))
```
Comment (gvardany) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py:662]: `PSD T0 Offset...`

File: Testing/Data/SystemTest/ILL/IN16B/ILLIN16B_BATS.nxs.md5 [hunk 4]
```diff
@@ -0,0 +1 @@
+5922f89b45b9f73becd769a2d23154e5
```
Comment (gvardany) [inline Testing/Data/SystemTest/ILL/IN16B/ILLIN16B_BATS.nxs.md5:1]: Is this file still needed?

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py [hunk 5]
```diff
@@ -592,26 +621,48 @@ def _reduce_bats(self, ws):
...
             else:
-                self._create_elastic_channel_ws(mtd[equator_epp_ws], mtd[ws].getRun())
-                RenameWorkspace(InputWorkspace=equator_epp_ws, OutputWorkspace=epp_ws)
+                single_det_ws = _make_name(ws, "sds")
+                ExtractSpectra(InputWorkspace=ws,
+                               OutputWorkspace=single_det_ws,
+                               StartWorkspaceIndex=N_TUBES*N_PIXELS_PER_TUBE + 1,
+                               EndWorkspaceIndex=N_TUBES*N_PIXELS_PER_TUBE + single_detectors)
+                FindEPP(InputWorkspace=single_det_ws, OutputWorkspace=epp_ws)
+                self._create_elastic_channel_ws(mtd[epp_ws], mtd[ws].getRun(), mtd[equator_epp_ws],
```
Comment (gvardany) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py:639]: This is a bit confusing, are you passing the epp of single detectors as the first argument, together with number of single detectors? If so, maybe clarify in the doxygen comments of _create_elastic...

**Review by gvardany — COMMENTED** [review]
Few comments on the code, will check with the new data shortly.

**Review by gvardany — COMMENTED** [review]
Tested with 2 new measurements, and it all worked as expected.
Just noticed an issue in case GroupDetectors=True.
This is not related to the grouping done for the peak finding.
This is the final grouping of the output.
The issue is that the default grouping file contains only 2 single detectors.
So if there are more, the rest are not included in the final result.
This cannot be added in the default grouping file, since in that case having 2 single detectors will fail due to unknown spectrum/detid for the 3rd. We have to think of something more dynamic but this can be done in a separate ticket. For this one, I'd only suggest to add one more system test with the new data, that has 3 single detectors (with GroupDetectors=False). The test should be similar to the existing one, i.e. use a center measurement and an offseted one.

**Comment by StephenSmith25** [comment]
Is it intended for this PR to make it into the 6.1 release?

**Comment by MathieuTillet** [comment]
Yes, it should be ready soon.

**Comment by gvardany** [comment]
Almost there. There is an issue still with FitEquatorialOnly, when having also an offset measurement:
```
config.appendDataSearchDir('/Volumes/illdata/211/in16b/internalUse/rawdata')
center, epp = IndirectILLEnergyTransfer(Run='318308', 
                                        PulseChopper='34',
                                        ElasticPeakFitting='FitEquatorialOnly',  
                                        GroupDetectors=False)
offset = IndirectILLEnergyTransfer(Run='315515:315521', 
                                   PulseChopper='34', 
                                   GroupDetectors=False,
                                   InputElasticChannelWorkspace=epp)
```
Take a look into this and turn this to a system test to cover also the offset measurement where this EPP table is actually used.
Ideally add 2 tests, one for each ElasticPeakFitting method.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py (hunk 2 above)
Comment (gvardany) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py:533]: This feature is not urgent, so I created a separate issue to address it later: https://github.com/mantidproject/mantid/issues/31399

**Review by gvardany — DISMISSED** [review]
Thanks for adding the test, looks good.

**Comment by gemmaguest** [comment]
@MathieuTillet Could you change the base branch to `release-next` if this is for the release?

**Review by gvardany — APPROVED** [review]
